File: python/src/yolo26_kit/core/letterbox.py
"""Letterbox coordinate utilities. See spec/decode.md Algorithm E."""
from __future__ import annotations

import numpy as np


def letterbox_unmap(
    boxes: np.ndarray,
    orig_size: tuple[int, int],   # (W_orig, H_orig)
    lb_size: tuple[int, int],     # (W_lb, H_lb)
    scale: float,
    pad: tuple[float, float],     # (pad_x, pad_y)
) -> np.ndarray:
    arr = np.asarray(boxes, dtype=np.float32).reshape(-1, 4)
    pad_x, pad_y = pad
    out = np.empty_like(arr)
    out[:, 0] = (arr[:, 0] - pad_x) / scale
    out[:, 1] = (arr[:, 1] - pad_y) / scale
    out[:, 2] = (arr[:, 2] - pad_x) / scale
    out[:, 3] = (arr[:, 3] - pad_y) / scale
    W, H = orig_size  # noqa: N806 (math convention)
    np.clip(out[:, 0], 0.0, W, out=out[:, 0])
    np.clip(out[:, 1], 0.0, H, out=out[:, 1])
    np.clip(out[:, 2], 0.0, W, out=out[:, 2])
    np.clip(out[:, 3], 0.0, H, out=out[:, 3])
    return out

File: js/src/core/letterbox.ts
export function letterboxUnmap(
  boxes: Float32Array,
  origSize: readonly [number, number],
  _lbSize: readonly [number, number],
  scale: number,
  pad: readonly [number, number],
): Float32Array {
  const W = origSize[0];
  const H = origSize[1];
  const padX = pad[0];
  const padY = pad[1];
  const N = boxes.length / 4;
  const out = new Float32Array(boxes.length);
  for (let i = 0; i < N; i++) {
    const o = i * 4;
    const x1 = ((boxes[o] as number) - padX) / scale;
    const y1 = ((boxes[o + 1] as number) - padY) / scale;
    const x2 = ((boxes[o + 2] as number) - padX) / scale;
    const y2 = ((boxes[o + 3] as number) - padY) / scale;
    out[o] = Math.max(0, Math.min(W, x1));
    out[o + 1] = Math.max(0, Math.min(H, y1));
    out[o + 2] = Math.max(0, Math.min(W, x2));
    out[o + 3] = Math.max(0, Math.min(H, y2));
  }
  return out;
}
